### ISSUE TITLE
Replace SAM with SAM 2

### ISSUE THREAD

**Issue by RHogan615** [cvat-ai/cvat#8231]
### Actions before raising this issue

- [X] I searched the existing issues and did not find anything similar.
- [X] I read/searched [the docs](https://docs.cvat.ai/docs/)

### Is your feature request related to a problem? Please describe.

While SAM is a good model, it would be nice to have the 'latest and greatest' capabilities.

### Describe the solution you'd like

Replace the current SAM model with SAM 2

### Describe alternatives you've considered

Replace the SAM implementation with the newest version

**Comment by medphisiker** [comment]
I wanted to ask you to please not replace SAM with SAM 2. Instead, I would like to request that you add another new model, SAM 2, to the existing models =)

**Comment by RHogan615** [comment]
> I wanted to ask you to please not replace SAM with SAM 2. Instead, I would like to request that you add another new model, SAM 2, to the existing models =)

Fair point

**Comment by HanClinto** [comment]
Especially for the tracking capabilities of SAM2, it feels like this would be a powerful addition for marking an object once and tracking it through a whole video.

**Comment by HanClinto** [comment]
Because SAM 2 has so many new capabilities, it feels like it can function both as an interactor as well as a tracker, and I'm not entirely sure which "bucket" to put it into to really take advantage of its full capabilities within CVAT.

**Comment by nmanovic** [comment]
Hi, we have added SAM2 on SaaS (https://app.cvat.ai/) and for Enterprise customers: https://www.cvat.ai/post/meta-segment-anything-model-v2-is-now-available-in-cvat-ai

We are working to add tracking capabilities of SAM2 into CVAT. Thanks for your comments.

**Comment by Youho99** [comment]
Another possible and very useful feature with models like SAM and SAM2 would be precision annotation in bounding boxes.

The idea is to make an imprecise bounding box around the object to be annotated. The bounding box is sent to the SAM or SAM2 model, which segments the main object from the bounding box it receives. Finally, the precise bounding box is recreated by taking the extremum coordinates at the top, left, bottom, right.

This would allow very quick and precise annotating, without having to zoom in on the image (very useful for precise annotation of small objects for example).

In my free time, I made a python script using this logic with SAM to make precision annotation, taking as input an annotation json (COCO format I think) and which output a json in the same format, with the precise bounding boxes recalculated.

I could make it available to you if necessary.

**Comment by grzleadams** [comment]
> Hi, we have added SAM2 on SaaS (https://app.cvat.ai/) and for Enterprise customers: https://www.cvat.ai/post/meta-segment-anything-model-v2-is-now-available-in-cvat-ai
> 
> We are working to add tracking capabilities of SAM2 into CVAT. Thanks for your comments.

Is there any guidance yet on how to deploy it for self-hosted CVAT? I don't see a [nuclio config](https://github.com/cvat-ai/cvat/blob/develop/serverless/pytorch/facebookresearch) for it.

**Comment by Youho99** [comment]
> > Hi, we have added SAM2 on SaaS (https://app.cvat.ai/) and for Enterprise customers: https://www.cvat.ai/post/meta-segment-anything-model-v2-is-now-available-in-cvat-ai
> > We are working to add tracking capabilities of SAM2 into CVAT. Thanks for your comments.
> 
> Is there any guidance yet on how to deploy SAM2 as a model for self-hosted CVAT? I don't see a [nuclio config](https://github.com/cvat-ai/cvat/blob/develop/serverless/pytorch/facebookresearch) for it.

+1

**Comment by alex-bronze-vision** [comment]
> > Hi, we have added SAM2 on SaaS (https://app.cvat.ai/) and for Enterprise customers: https://www.cvat.ai/post/meta-segment-anything-model-v2-is-now-available-in-cvat-ai
> > We are working to add tracking capabilities of SAM2 into CVAT. Thanks for your comments.
> 
> Is there any guidance yet on how to deploy SAM2 as a model for self-hosted CVAT? I don't see a [nuclio config](https://github.com/cvat-ai/cvat/blob/develop/serverless/pytorch/facebookresearch) for it.

+1 too

**Comment by gauravlochab** [comment]
> 2 on S

+1 too

**Comment by amrosado** [comment]
Are there plans to bring this feature to non-enterprise users?  I couldn't easily find a branch with this feature being worked, but I can see support of bounding box input added in the develop branch.  This feature would be great for a project I'm working on.

**Comment by earzamastsev** [comment]
There are approximate plans for the implementation of SAM2 tracking (next month, before the end of the year...)?

**Comment by nmanovic** [comment]
@earzamastsev , we want to add SAM2 tracking in the next quarter. It will be an Enterprise feature as well. Thus users of SaaS and our Enterprise customers will be able to use it for their projects.

@amrosado , we don't have such plans. Please use SAM instead. It has really great quality. The primary benefit of SAM2 is the video tracking capabilities and inference speed. Again, if you want to support development of the open-source project, just buy a subscription and use all our paid features. For now there are no any other hidden or extra costs. Such simple action allows us to fix bugs and add new features. Otherwise, community can contribute their source code. We will keep it inside one of our repos or point in our documentation.

**Comment by ryanalexmartin** [comment]
> just buy a subscription and use all our paid features. 
 
 This comes across as quite out-of-touch, and rather rude considering CVAT has over 252 contributors, and according to your website you have 10 employees.
  
 > Otherwise, community can contribute their source code. 
 
 Somebody contributed a SAM2 interactor that you guys still refuse to merge in...

**Comment by nmanovic** [comment]
@ryanalexmartin, thank you for your comments and for submitting an issue several years ago. We value every community member and contributor. All contributions from our community are currently open-source, and we have supported these contributions for years. You can always use the contributed SAM2. It is a separate directory with files for the Nulcio serverless framework. Could you clarify the problem? The PR hasn't been merged yet because it still has issues with the linter. (@azhavoro , could you please ping the author of the PR and ask to fix remaining issues?)

To keep CVAT alive and valuable for the open-source community, we develop it according to open-core principles. Most of the functionality remains open-source, but key features that help automate data annotation, quality control, and team management will be provided to customers who support the project. Any organization can pay us once to implement a feature for the open-source community: https://www.cvat.ai/post/request-features. This approach has helped introduce several important features like skeleton support, quality control, and webhooks.

P.S. Feel free to contribute any feature to CVAT to make it available for the open-source community. However, please keep in mind that you will need to support it over the years as well.

**Comment by jeanchristopheruel** [comment]
Lol lets start a CVAT2 official fork that integrate SAM2 tracking + full frontend encoding + decoding

**Comment by corkwing** [comment]
> @earzamastsev , we want to add SAM2 tracking in the next quarter. It will be an Enterprise feature as well. Thus users of SaaS and our Enterprise customers will be able to use it for their projects.

Is it possible to get an update on when SAM2 for tracking will be available in CVAT? Thanks!

**Comment by nmanovic** [comment]
@corkwing , it will be available to our Enterprise customers this year. We will think how to add it for SaaS customers. The model requires a lot of resources to track an object on multiple frames. It will not be available under MIT license in the nearest future. We are thinking on how to introduce it to our open-source community.

**Comment by eebowen** [comment]
> Lol lets start a CVAT2 official fork that integrate SAM2 tracking + full frontend encoding + decoding for better control over PRs competing with commercial features. Community > Commercial.

Has anyone started this repo?

**Comment by nmanovic** [comment]
SAM v2 for videos are available for CVAT Online (SaaS) customers: https://www.cvat.ai/resources/blog/sam2-ai-agent-tracking

**Comment by Kieaer** [comment]
Personally, I have 3080Ti pc and A6000 GPU x 8 server
I was looking for a program that supports video SAM2 labeling to use for AI labeling after using it all the time to create AI videos, and I found CVAT first.
I thought it would be better to fix the `label-studio-ml-backend` that doesn't work well rather than using CVAT, which only works online and can't utilize the my local GPU x 8 server.
.. or use cvat sam2 fork version 👀

**Comment by nmanovic** [comment]
@Kieaer , it can. SAMv2 for videos: https://www.cvat.ai/resources/blog/sam2-ai-agent-tracking
You can run it on your machine and connect to CVAT Online.

**Comment by Kieaer** [comment]
@nmanovic I've already seen this document, but it took too long for CVAT online to upload 2-3GB for each video data, 20-40 minutes each.
CVAT online was not suitable because the total video dataset was 1TB.
So I'm looking for a local to be SAM2 labeling :eyes:

**Comment by jneeven** [comment]
> Lol lets start a CVAT2 official fork that integrate SAM2 tracking + full frontend encoding + decoding for better control over PRs competing with commercial features. Community > Commercial.

@jeanchristopheruel I would be interested in this. I'm currently using https://github.com/cvat-ai/cvat/pull/8610 with custom finetuned models with great success. However, since my images are really (time-subsampled) video frames, I think incorporating the time dimension could really make labeling an order of magnitude more efficient. I'd love to work on something like this, but am not familiar with the CVAT source code at all. I've looked at the source code of that SAM2.1 plugin in the PR I link above, and more or less understand what it does, but what I don't understand is what CVAT code takes my mouse clicks, feeds them to that plugin, and parses the results. Beyond implementing the SAM2 encoder in the frontend, do you have any idea what kind of CVAT code would need to be modified to support tracking, and if so, could you please give a rough overview of the steps involved?

Many thanks!
